### ISSUE TITLE
Question: Sliding Window Module for Transformer3dSeg Object

### ISSUE THREAD

**Issue by jmarsil** [The-AI-Summer/self-attention-cv#2]
I was wondering whether or not you've implemented an example using the network in a 3d medical segmentation task and/or use case? If this network only exports the center slice of a patch then we would need a wrapper function to iterate through all patches in an image to get the final prediction for the entire volume. From the original paper, I assume they choose 10 patches at random from an image during training, but it's not too clear how they pieced everything together during testing.

Your thoughts on this would be greatly appreciated!

See:
https://github.com/The-AI-Summer/self-attention-cv/blob/33ddf020d2d9fb9c4a4a3b9938383dc9b7405d8c/self_attention_cv/Transformer3Dsegmentation/tranf3Dseg.py#L10

**Comment by black0017** [comment]
Hello, I still haven't figured that out yet. Would you like to help me out? 

My current problem is that the input is a 3d volume of subvol_dim^3 and the output is a volume of (subvol_dim//patch_dim)^3. 

![image](https://user-images.githubusercontent.com/13345132/109868702-41f5c180-7c68-11eb-84f2-b1c9f7352ba9.png)

You can take 3D sub-volumes from a large 3D volume using [torch.io](https://github.com/fepegar/torchio) which I originally planned to use.

**Comment by jmarsil** [comment]
Sure I'd love to. I can dive into it and try to come up with a solution.

**Comment by JavisPeng** [comment]
@black0017 I have the same question, in my view ,the output size should be 8*8*8*num_classes, is the right ?

**Comment by black0017** [comment]
@JavisPeng @jmarsil just found a colab notebook that I have been experimenting with torch.io https://colab.research.google.com/drive/19q84fPhPhGmX36SMdrWfoJBN7k6eoSR2?usp=sharing

what @JavisPeng  mentions is for the inference part

@JavisPeng correct https://github.com/The-AI-Summer/self-attention-cv/blob/main/self_attention_cv/Transformer3Dsegmentation/tranf3Dseg.py#L96
I do that already but I didn't mention the number of classes in the prev. comment.

Now that I think about it training would be ok. You just have to use only the 8x8x8 part from the sub-sampled label map (see the notebook) that is sampled from torch.io sampler. 

here is a snippet from the notebook:
```python
for epoch_index in range(num_epochs):
  for patches_batch in patches_loader:
    
    input_t1 = patches_batch['t1'][tio.DATA]  
    input_t2 = patches_batch['t2'][tio.DATA]  

    # for the 3d transformer-based approach the target must be cropped again to the desired size
    targets = patches_batch['label'][tio.DATA]  
    
    logits = model(input_t1) # 8x8x8 the 3d transformer-based approach
```

The question comes for inference. Any idea how to change the inference part to support what the authors do in their work?
I mean from 24x24x24 -> 8x8x8 seg sound kind of buggy for inference.

**Comment by black0017** [comment]
Hello,

I have performed a minimal example with torch.io in the notebook. (train + inference)

Not 100% about the inference :)

Check it out here:
https://colab.research.google.com/drive/19q84fPhPhGmX36SMdrWfoJBN7k6eoSR2?usp=sharing

I am closing the issue.

Let me know if it works out in your experiments.

I will add a folder with notebooks.
